spring-pulsar/src/test/java/org/springframework/pulsar/listener/DefaultPulsarMessageListenerContainerTests.java

```diff
@@ -87,6 +94,76 @@ void basicDefaultConsumer() throws Exception {
 		pulsarClient.close();
 	}
 
+	@Test
+	void containerPauseAndResumeFeatureUsingWaitAndNotify() throws Exception {
+		HashSet<String> topics = new HashSet<>();
+		topics.add("containerPauseResumeWaitNotify-topic");
+		Map<String, Object> config = Map.of("topicNames", topics, "subscriptionName",
+				"containerPauseResumeWaitNotify-sub");
+		PulsarClient pulsarClient = PulsarClient.builder().serviceUrl(PulsarTestContainerSupport.getPulsarBrokerUrl())
+				.build();
+		DefaultPulsarConsumerFactory<String> pulsarConsumerFactory = new DefaultPulsarConsumerFactory<>(pulsarClient,
+				config);
+		PulsarContainerProperties pulsarContainerProperties = new PulsarContainerProperties();
+		pulsarContainerProperties.setMessageListener((PulsarRecordMessageListener<?>) (consumer, msg) -> {
+		});
+		pulsarContainerProperties.setSchema(Schema.STRING);
+		DefaultPulsarMessageListenerContainer<String> container = new DefaultPulsarMessageListenerContainer<>(
+				pulsarConsumerFactory, pulsarContainerProperties);
+
+		Lock reentrantLock = new ReentrantLock();
+		Condition lockCondition = reentrantLock.newCondition();
+
+		Lock spyLock = spy(reentrantLock);
+		Condition spyCondition = spy(lockCondition);
+
+		ReflectionTestUtils.setField(container, "lockOnPause", spyLock);
+		ReflectionTestUtils.setField(container, "pausedCondition", spyCondition);
+
+		CountDownLatch latchOnLockInvocation = new CountDownLatch(2);
+		CountDownLatch latchOnUnlockInvocation = new CountDownLatch(2);
+		CountDownLatch latchOnAwaitInvocation = new CountDownLatch(1);
+		CountDownLatch latchOnSignalInvocation = new CountDownLatch(1);
+
+		doAnswer(invocation -> {
+			latchOnLockInvocation.countDown();
+			return invocation.callRealMethod();
+		}).when(spyLock).lock();
+
+		doAnswer(invocation -> {
+			latchOnUnlockInvocation.countDown();
+			return invocation.callRealMethod();
+		}).when(spyLock).unlock();
+
+		doAnswer(invocation -> {
+			latchOnAwaitInvocation.countDown();
+			return invocation.callRealMethod();
+		}).when(spyCondition).await();
+
+		doAnswer(invocation -> {
+			latchOnSignalInvocation.countDown();
+			return invocation.callRealMethod();
+		}).when(spyCondition).signal();
+
+		container.start();
+
+		container.pause();
+
+		Awaitility.await().until(container::isPaused);
+
+		container.resume();
+
+		Awaitility.await().until(() -> !container.isPaused());
+
+		assertThat(latchOnLockInvocation.await(10, TimeUnit.SECONDS)).isTrue();
```

Comment:
This test verifies the use of the locks in the current impl really well. I wonder if we should also add an IT that sends/receives messages and make sure it behaves as expected?

Reply:
Ok, I will think about that (maybe outside of this PR).